engine.go

```diff
@@ -19,13 +18,16 @@ import (
 
 const parallelism = 5
 
+type WriteHandler *func(ctx context.Context, block *types.Block) error
```

Comment:
I think we can make this required (not a pointer) and pass in the base implementation